core/src/main/java/io/questdb/cairo/map/FastMap.java

```diff
@@ -279,17 +296,24 @@ public void reopen() {
 
     @Override
     public void restoreInitialCapacity() {
-        heapStart = kPos = Unsafe.realloc(heapStart, heapLimit - heapStart, capacity = initialPageSize, heapMemoryTag);
-        heapLimit = heapStart + initialPageSize;
-        keyCapacity = (int) (initialKeyCapacity / loadFactor);
-        keyCapacity = keyCapacity < MIN_INITIAL_CAPACITY ? MIN_INITIAL_CAPACITY : Numbers.ceilPow2(keyCapacity);
-        mask = keyCapacity - 1;
-        free = (int) (keyCapacity * loadFactor);
-        offsets.resetCapacity();
-        offsets.setCapacity((long) keyCapacity << 1);
-        offsets.setPos((long) keyCapacity << 1);
-        offsets.zero(0);
-        nResizes = 0;
+        if (capacity != initialPageSize || keyCapacity != offsets.getCapacity()) {
+            heapStart = kPos = Unsafe.realloc(heapStart, heapLimit - heapStart, capacity = initialPageSize, heapMemoryTag);
+            heapLimit = heapStart + initialPageSize;
+            keyCapacity = (int) (initialKeyCapacity / loadFactor);
+            keyCapacity = keyCapacity < MIN_INITIAL_CAPACITY ? MIN_INITIAL_CAPACITY : Numbers.ceilPow2(keyCapacity);
+            mask = keyCapacity - 1;
+            free = (int) (keyCapacity * loadFactor);
+            offsets.resetCapacity();
+            offsets.setCapacity((long) keyCapacity << 1);
+            offsets.setPos((long) keyCapacity << 1);
+            offsets.zero(0);
+            nResizes = 0;
+        }
+    }
+
+    @Override
+    public void setKeyCapacity(int newKeyCapacity) {
+        rehash(Numbers.ceilPow2((int) (newKeyCapacity / loadFactor)));
```

Comment:
nit: safer to verify the capacity first
```suggestion
        if (newKeyCapacity > keyCapacity) {
            rehash(Numbers.ceilPow2((int) (newKeyCapacity / loadFactor)));
        }
```